tensor/datafs/dir.go

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) 2024, Cogent Core. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package datafs
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"path"
+	"slices"
+	"sort"
+
+	"golang.org/x/exp/maps"
+)
+
+// NewDir returns a new datafs directory with given name.
+// if parent != nil and a directory, this dir is added to it.
+// if name is empty, then it is set to "/", the root directory.
+// Names must be unique within a directory.
+func NewDir(name string, parent ...*Data) (*Data, error) {
+	if name == "" {
+		name = "/"
+	}
+	var par *Data
+	if len(parent) == 1 {
+		par = parent[0]
+	}
+	d, err := NewData(par, name)
+	d.Value = make(map[string]*Data)
+	return d, err
+}
+
+// Item returns data item in given directory by name.
+// This is for fast access and direct usage of known
+// items, and it will crash if item is not found or
+// this data is not a directory.
+func (d *Data) Item(name string) *Data {
+	fm := d.filemap()
+	return fm[name]
+}
+
+// Items returns data items in given directory by name.
+// error reports any items not found, or if not a directory.
+func (d *Data) Items(names ...string) ([]*Data, error) {
+	if err := d.mustDir("Items", ""); err != nil {
+		return nil, err
+	}
+	fm := d.filemap()
+	var errs []error
+	var its []*Data
+	for _, nm := range names {
+		dt := fm[nm]
+		if dt != nil {
+			its = append(its, dt)
+		} else {
+			err := fmt.Errorf("datafs Dir %q item not found: %q", d.Path(), nm)
+			errs = append(errs, err)
+		}
+	}
+	return its, errors.Join(errs...)
+}
+
+// ItemsFunc returns data items in given directory
+// filtered by given function, in alpha order.
+// If func is nil, all items are returned.
+// Any directories within this directory are returned,
+// unless specifically filtered.
+func (d *Data) ItemsFunc(fun func(item *Data) bool) []*Data {
+	if err := d.mustDir("ItemsFunc", ""); err != nil {
+		return nil
+	}
+	fm := d.filemap()
+	names := d.DirNamesAlpha()
+	var its []*Data
+	for _, nm := range names {
+		dt := fm[nm]
+		if fun != nil && !fun(dt) {
+			continue
+		}
+		its = append(its, dt)
+	}
+	return its
+}
+
+// ItemsByTimeFunc returns data items in given directory
+// filtered by given function, in time order (i.e., order added).
+// If func is nil, all items are returned.
+// Any directories within this directory are returned,
+// unless specifically filtered.
+func (d *Data) ItemsByTimeFunc(fun func(item *Data) bool) []*Data {
+	if err := d.mustDir("ItemsByTimeFunc", ""); err != nil {
+		return nil
+	}
+	fm := d.filemap()
+	names := d.DirNamesByTime()
+	var its []*Data
+	for _, nm := range names {
+		dt := fm[nm]
+		if fun != nil && !fun(dt) {
+			continue
+		}
+		its = append(its, dt)
+	}
+	return its
+}
+
+// FlatItemsFunc returns all "leaf" (non directory) data items
```

Comment:
Should you use the terminology `file` instead of or in addition to `leaf`?

Reply:
no. file includes directory from the Go fs impl.